lib/gadget/gadget.vala

```diff
@@ -100,12 +104,30 @@ namespace Frida.Gadget {
 			}
 			resume ();
 		} else {
-			var s = new Server (jit_enabled, gadget_range);
+
+			string listen_uri;
 			try {
-				yield s.start ();
-				server = s;
-				log_info ("Listening on TCP port 27042");
-			} catch (Error e) {
+				string? listen_address = GLib.Environment.get_variable ("FRIDA_GADGET_LISTEN_ADDRESS");
+				var raw_address = (listen_address != null) ? listen_address : DEFAULT_LISTEN_ADDRESS;
+				var socket_address = NetworkAddress.parse (raw_address, DEFAULT_LISTEN_PORT).enumerate ().next ();
+				if (socket_address is InetSocketAddress) {
+					var inet_socket_address = socket_address as InetSocketAddress;
+					var inet_address = inet_socket_address.get_address ();
+					var family = (inet_address.get_family () == SocketFamily.IPV6) ? "ipv6" : "ipv4";
+					string address = inet_address.to_string ();
+					uint16 port = inet_socket_address.get_port ();
+					listen_uri = "tcp:family=%s,host=%s,port=%hu".printf (family, address, port);
```

Comment:
This feels like it belongs in a separate function, or at least in a separate try/catch block above. But I'd lean towards a function, perhaps something like:

```Vala
string listen_address;
uint16 listen_port;
string listen_uri;
if (!try_get_listen_address (out listen_address, out listen_port, out listen_uri)) {
	log_error ("Invalid listen address");
	return;
}

try {
	var s = new Server (listen_uri, jit_enabled, gadget_range);
	yield s.start ();

	server = s;

	log_info ("Listening on %s TCP port %hu", listen_address, listen_port);
} catch (GLib.Error e) {
	log_error ("Failed to start: %s", e.message);
}
```

and then:

```Vala
private bool try_get_listen_address (out string address, out uint16 port, out string uri) {
}
```